### PR TITLE
RFC: Refactor log-extractor to be modular, generic regarding input files and testable.

### PR DESCRIPTION
**Pull request by KKoukiou**
This patch introduces the following changes:

CORE:

  * Add the ability to parse logs from various sources. (url, zip, directory)
  * Introduce single parameter for input called source. Remove all other related params.
  * Refactor script to not unpack all files to utilize better disk space.

TESTING:

  * Create test folder to prepare ground for future tests.
  * Add simple tests for LZMAFile class.
  * Add tox.ini file.
  * Add .travis.yaml file

DEVELOPER:

  * Rename log_extractor.py to extractor.py: module name should not be same
	as package name if we want to import from it.
  * Change setup.py to use pbr [1]
  * Separate requirements file into requuirements.txt to help with reusability.

[1]: https://docs.openstack.org/pbr/latest/

### CODE REVIEW

**Comment by KKoukiou** [comment]
As I mention in subject in RFC, please let me know what you think.

**Comment by KKoukiou** [comment]
I forgot to mention that this patch also prepares the ground for semantic versioning so let me know if we can start with that as well.

File: README.md [hunk 1]
```diff
@@ -1,17 +1,35 @@
 # log-extractor
-ART log extractor, download a job artifacts, unpack all archives and parse relevant logs.
+When running lots of tests, it is easy for logs to get huge and trying to
+search through them to find the data that is relevant to the portion of time
+a specific test was running for is tedious, even with time stamping.
+log-extractor rei-structures the logs, creating per testcase log files.
 
 # Install
 `sudo pip install . -U`
```
Comment (myakove) [inline README.md:8]: 'pip install . -U --user' is better.

File: README.md [hunk 2]
```diff
@@ -1,17 +1,35 @@
 # log-extractor
-ART log extractor, download a job artifacts, unpack all archives and parse relevant logs.
+When running lots of tests, it is easy for logs to get huge and trying to
+search through them to find the data that is relevant to the portion of time
+a specific test was running for is tedious, even with time stamping.
+log-extractor rei-structures the logs, creating per testcase log files.
```
Comment (kobihk) [inline README.md:5]: rei-structures ??

File: log_extractor/extractor.py [hunk 3]
```diff
@@ -270,15 +216,58 @@ def _create_test_dir(self, line):
...
-    def parse_art_logs(self, team=None):
+    def unpack_relevant_remote_logs(self, dst, source_object):
+        """"
+        Unpacks tar.gz files containing remote logs in case they are
+        inside archive.
+
+        Args:
+            dst (str): Destination path to store the extracted files
+            source_object (str): Object containing log directory information
+
```
Comment (cynepco3hahue) [inline log_extractor/extractor.py:227]: Remove Return

File: log_extractor/extractor.py [hunk 4]
```diff
@@ -270,15 +216,58 @@ def _create_test_dir(self, line):
...
 
-    def parse_art_logs(self, team=None):
+    def unpack_relevant_remote_logs(self, dst, source_object):
+        """"
+        Unpacks tar.gz files containing remote logs in case they are
+        inside archive.
+
+        Args:
+            dst (str): Destination path to store the extracted files
+            source_object (str): Object containing log directory information
```
Comment (cynepco3hahue) [inline log_extractor/extractor.py:226]: Change source_object type

File: log_extractor/extractor.py [hunk 5]
```diff
@@ -270,15 +216,58 @@ def _create_test_dir(self, line):
...
+
+        Returns:
+
+        """
```
Comment (cynepco3hahue) [inline log_extractor/extractor.py:230]: It looks fine, but I still want to test it at least on 4.1 and 4.2, also we still have runs for 3.6 and 4.0 that archive and logs names look different

**Review by cynepco3hahue — CHANGES_REQUESTED** [review]
The patch looks fine, but it needs a lot of testing.
I tried to run 
`python log_extractor/extractor.py --source https://rhv-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/4.2_dev/job/rhv-4.2-ge-runner-compute/273/`
and received 
```
Traceback (most recent call last):
  File "log_extractor/extractor.py", line 545, in <module>
    run()
  File "/usr/lib64/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/lib64/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib64/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "log_extractor/extractor.py", line 505, in run
    source_type = helper.identify_source_type(source)
  File "/home/alukiano/PycharmProjects/log-extractor/log_extractor/helper.py", line 51, in identify_source_type
    f = urlopen(source)
  File "/usr/lib64/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib64/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib64/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib64/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 1243, in https_open
    context=self._context)
  File "/usr/lib64/python2.7/urllib2.py", line 1200, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:661)>
```

File: log_extractor/extractor.py (hunk 5 above)
Comment (KKoukiou) [reply log_extractor/extractor.py:230]: We don't run 4.0 anymore, but you 're right about 3.6.. I 'll have to make it working.

**Comment by KKoukiou** [comment]
Regarding the certificate Error it's because you probably don't have the certificate for Jenkins in your machine. In your code, you don't test the certificate, but in my I do.
To my opinion everybody should use certificates to connect to Jenkins, but it's up to you. If you prefer to not use it, I can add the option in the urlopen

**Comment by KKoukiou** [comment]
Regarding the certificate Error it's because you probably don't have the certificate for Jenkins in your machine. In your code, you don't test the certificate, but in my I do.
To my opinion everybody should use certificates to connect to Jenkins, but it's up to you. If you prefer to not use it, I can add the option in the urlopen

**Comment by KKoukiou** [comment]
@cynepco3hahue So, I was thinking.. Do we really need to make the script more complicated to support the old logs format? We can tag it and we will know which tag is able to run for 3.6.. What do you think?

File: log_extractor/constants.py [hunk 6]
```diff
@@ -5,23 +5,31 @@
...
 """
 
-import os
-import user
-
 JOB_ARTIFACT = 'artifact'
 JOB_ARTIFACT_ZIP = '{0}/*zip*/archive.zip'.format(JOB_ARTIFACT)
-ARCHIVE_EXTENSIONS = ['.gz', '.zip', '.xz', '.tar', '.bz2']
-TEAMS = ['compute', 'networking', 'storage', 'system', 'coresystem', 'sla', 'virt']
+ARTIFACT_ZIP_NAME = "archive.zip"
```
Comment (cynepco3hahue) [inline log_extractor/constants.py:10]: I think it must be `artifact.zip`

**Review by cynepco3hahue — APPROVED** [review]
Ok I found some other problems
- no nested extraction for `engine.log.gz`
- incorrect extract directory when you specify archive as a source, I think it must be the same directory where the archive is

But I will merge it anyway because I already prepared a patch that will all this stuff and some additional problems with engine log parsing :smile:

**Pull request merged**